### ISSUE TITLE
testnet node crashed on IncrementNoteWitnesses

### ISSUE THREAD

**Issue by ebfull** [zcash/zcash#6004]
My testnet node was reindexing (I think reindex had completed and now threadnotifywallets was working) and aborted:

```
zcashd: wallet/wallet.cpp:2639: void CWallet::IncrementNoteWitnesses(const Consensus::Params &, const CBlockIndex *, const CBlock *, MerkleFrontiers &, bool): Assertion `orchardWallet.CheckpointNoteCommitmentTree(pindex->nHeight)' failed.
Aborted (core dumped)
```

Last log entry was

```
2022-06-08T20:42:12.308168Z ERROR rustzcash::wallet: Expected checkpoint height 1864618, given 1842420
```

which is a pretty huge disparity in block heights

**Comment by defuse** [comment]
Same thing happened to me, but on mainnet. The log ends with:

```
2023-01-30T10:37:11.172276Z  INFO main: UpdateTip: new best hash=00000000004c93b6174b79d73cb95081fd6575067375be23bc71062530210fec height=1789636 bits=469858871 log2_work=59.358472 tx=11491053 date=2022-08-29 05:01:16 progress=0.853724 cache=47.7MiB(6545tx)
2023-01-30T10:37:11.561955Z  INFO main: UpdateTip: new best hash=00000000010ed2b6792ac3fae3963d631317d67cdec3db8cb9f32eb2ce623ceb height=1789637 bits=469860485 log2_work=59.358474 tx=11491059 date=2022-08-29 05:01:50 progress=0.853724 cache=47.7MiB(6547tx)
2023-01-30T10:37:11.817145Z  INFO main: UpdateTip: new best hash=00000000011b0af8bfb10606eee28e28503921743040e0bdb39de4ac3b768feb height=1789638 bits=469860598 log2_work=59.358475 tx=11491062 date=2022-08-29 05:01:54 progress=0.853725 cache=47.7MiB(6548tx)
2023-01-30T10:37:12.092166Z  INFO main: UpdateTip: new best hash=0000000001487ec1dd800336656e9472fedbe1382abf9b550d50b8c99f6112c2 height=1789639 bits=469858164 log2_work=59.358477 tx=11491065 date=2022-08-29 05:02:41 progress=0.853725 cache=47.8MiB(6549tx)
2023-01-30T10:37:12.356453Z  INFO main: UpdateTip: new best hash=0000000000188336214e034f87ecae14c085340930b33092c6baf450f7bcebaf height=1789640 bits=469856877 log2_work=59.358478 tx=11491070 date=2022-08-29 05:04:16 progress=0.853725 cache=47.8MiB(6551tx)
2023-01-30T10:37:12.358394Z ERROR rustzcash::wallet: Expected checkpoint height 1965726, given 1687104
```

This node had previously finished syncing and I was reindexing after adding `txindex=1` and `insightexplorer=1` into my zcash.conf. Prior to the reindex while it was fully synced it had crashed running out of memory, in case that might have caused some kind of corruption. While it was reindexing, it seemed to start re-downloading all the blocks (showing that yellow progress bar where it has the headers but not the blocks yet). Then this crash happened in that block re-downloading process.

**Comment by str4d** [comment]
I hit this same exception while reindexing on `v5.4.0-rc3` to test the transparent pool tracking. I run my node with `txindex=1` and `prometheusport` set, and just restarted with `zcashd -reindex`.

Last metrics screen output:
```
           Block height | 1965752
           Next upgrade | Unknown
            Connections | 8
  Network solution rate | 12.235 GSol/s


Since starting this node 1 days, 19 hours, 51 minutes, 52 seconds ago:
- You have validated 20493755 transactions!

[Press Ctrl+C to exit] [Set 'showmetrics=0' to hide]
zcashd: wallet/wallet.cpp:2857: void CWallet::IncrementNoteWitnesses(const Consensus::Params &, const CBlockIndex *, const CBlock *, MerkleFrontiers &, bool): Assertion `orchardWallet.CheckpointNoteCommitmentTree(pindex->nHeight)' failed.
Aborted (core dumped)
```


`debug.log` ending:
```
2023-01-29T19:42:14.303943Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=0000000000c06e0e0d755638cae058e5586647c5bf5acfabadc2b30f3137419c height=1965748 bits=469844242 log2_work=59.599141 tx=12208217 date=2023-01-29 19:42:01 progress=1.000000 cache=226.7MiB(8963tx)
2023-01-29T19:43:33.884248Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=00000000010455c52071809fb580c1b4c14d2541ab2aafd39a78e949d8743a36 height=1965749 bits=469841389 log2_work=59.599143 tx=12208222 date=2023-01-29 19:42:59 progress=0.999999 cache=242.7MiB(8964tx)
2023-01-29T19:45:18.159323Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=0000000000ef3559f33b09a50e573178531618c4178f645f9582113d89aec088 height=1965750 bits=469840828 log2_work=59.599144 tx=12208229 date=2023-01-29 19:45:02 progress=1.000000 cache=264.0MiB(8968tx)
2023-01-29T19:49:45.493740Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=000000000079ce47dcf579c97dfc750edd9a67c08f35cc44a0ef584dc1248db5 height=1965751 bits=469843521 log2_work=59.599146 tx=12208240 date=2023-01-29 19:49:20 progress=1.000000 cache=98.1MiB(27tx)
2023-01-29T19:50:03.406837Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=00000000002c2c035a8c82e1a43c9494348bb93c49d9fc165cf26fc20cc6987e height=1965752 bits=469842960 log2_work=59.599147 tx=12208245 date=2023-01-29 19:49:44 progress=1.000000 cache=100.4MiB(30tx)
2023-01-29T19:51:54.062864Z ERROR rustzcash::wallet: Expected checkpoint height 1960017, given 1687104
```

**Comment by str4d** [comment]
We've identified the problem and fixed it in #6387.

I've restarted my node on a build of that PR, without `-reindex` (so the changes in the PR are not triggered), and my node started rescanning:
```
2023-01-31T00:34:16.732664Z  INFO Init: main: CWallet::InitLoadWallet(): Rescanning last 283749 blocks (from block 1682001)...
2023-01-31T00:35:16.057793Z  INFO Init: main: Still rescanning. At block 1707011. Progress=0.793151
2023-01-31T00:36:16.126616Z  INFO Init: main: Still rescanning. At block 1708239. Progress=0.794462
```

This exactly matches our hypothesis for what would happen if you interrupted the reindex manually with a restart, and indeed my node is able to scan past NU5 activation. So this bug AFAICT was recoverable without #6387, but should no longer occur.
